File: packages/quran-engine-go/word_by_word.go
package quranengine

// Word by word: what each word of an ayah means, and how it is said.
//
// Two layers over the SAME tokens — the English gloss and a Latin transliteration — where the
// tokens are the ayah's own whitespace-separated words. Split the ayah and index straight in; the
// alignment against a corpus that tokenizes ~200 ayahs differently was done once, at build time.
//
// A token with no word of its own (the ۞ ornament, the tail of a word the corpus writes as two)
// carries "" in both layers — show nothing for it rather than a neighbour's meaning.
//
// See ../../docs/12-word-by-word.md.

import (
	"strconv"
	"strings"
)

// WordByWordPack is data/word-by-word.json: surah id -> ayahs in id order -> one entry per token.
type WordByWordPack struct {
	English         map[string][][]string `json:"english"`
	Transliteration map[string][][]string `json:"transliteration"`
}

// GlossedWord is one word of an ayah, in reading order.
type GlossedWord struct {
	// Position is the 1-based index of the word in the ayah.
	Position int
	// Arabic is the ayah's own token.
	Arabic string
	// English is the gloss, "" when the token has none.
	English         string
	Transliteration string
}

// GlossHit is a hit from the word-level gloss search.
type GlossHit struct {
	Surah           int
	Ayah            int
	Position        int
	English         string
	Transliteration string
}

// WordByWordLoaded reports whether a pack is loaded at all — cheap enough to gate UI on.
func (e *Engine) WordByWordLoaded() bool {
	return len(e.wordByWord.English) > 0
}

// Words returns every word of an ayah, in reading order.
func (e *Engine) Words(surahID, ayahID int) []GlossedWord {
	english, ok := e.Glosses(surahID, ayahID)
	if !ok {
		return nil
	}
	latin, _ := e.Transliterations(surahID, ayahID)
	var tokens []string
	if ayah := e.Ayah(surahID, ayahID); ayah != nil {
		tokens = strings.Fields(ayah.TextArabic)
	}

	out := make([]GlossedWord, 0, len(english))
	for i, gloss := range english {
		word := GlossedWord{Position: i + 1, English: gloss}
		if i < len(tokens) {
			word.Arabic = tokens[i]
		}
		if i < len(latin) {
			word.Transliteration = latin[i]
		}
		out = append(out, word)
	}
	return out
}

// Word returns one word by its 1-based position.
func (e *Engine) Word(surahID, ayahID, position int) *GlossedWord {
	words := e.Words(surahID, ayahID)
	if position < 1 || position > len(words) {
		return nil
	}
	return &words[position-1]
}

// Glosses returns the English layer for an ayah. The second result is false when the pack has no
// row for it, which is not the same as an ayah every one of whose tokens is unglossed.
func (e *Engine) Glosses(surahID, ayahID int) ([]string, bool) {
	return layerRow(e.wordByWord.English, surahID, ayahID)
}

// Transliterations returns the Latin layer, aligned token for token with Glosses.
func (e *Engine) Transliterations(surahID, ayahID int) ([]string, bool) {
	return layerRow(e.wordByWord.Transliteration, surahID, ayahID)
}

// FindGloss lists ayahs containing a word whose gloss carries term — a word-level English search,
// which finds ayahs a translation search misses because no translator used that phrasing.
func (e *Engine) FindGloss(term string, limit int) []GlossHit {
	needle := strings.ToLower(strings.TrimSpace(term))
	if needle == "" {
		return nil
	}
	var out []GlossHit
	// Mushaf order, so the result is stable across runs (map iteration is not).
	for i := range e.surahs {
		surah := &e.surahs[i]
		rows, ok := e.wordByWord.English[strconv.Itoa(surah.ID)]
		if !ok {
			continue
		}
		for ayahIndex, glosses := range rows {
			for index, gloss := range glosses {
				if !strings.Contains(strings.ToLower(gloss), needle) {
					continue
				}
				latin := ""
				if latinRows, ok := e.wordByWord.Transliteration[strconv.Itoa(surah.ID)]; ok &&
					ayahIndex < len(latinRows) && index < len(latinRows[ayahIndex]) {
					latin = latinRows[ayahIndex][index]
				}
				out = append(out, GlossHit{
					Surah:           surah.ID,
					Ayah:            ayahIndex + 1,
					Position:        index + 1,
					English:         gloss,
					Transliteration: latin,
				})
				if limit > 0 && len(out) >= limit {
					return out
				}
			}
		}
	}
	return out
}

func layerRow(layer map[string][][]string, surahID, ayahID int) ([]string, bool) {
	rows, ok := layer[strconv.Itoa(surahID)]
	if !ok || ayahID < 1 || ayahID > len(rows) {
		return nil, false
	}
	return rows[ayahID-1], true
}

File: packages/quran-engine-rust/src/word_by_word.rs
//! Word by word: what each word of an ayah means, and how it is said.
//!
//! Two layers over the SAME tokens — the English gloss and a Latin transliteration — where the
//! tokens are the ayah's own whitespace-separated words. Split the ayah and index straight in; the
//! alignment against a corpus that tokenizes ~200 ayahs differently was done once, at build time.
//!
//! A token with no word of its own (the ۞ ornament, the tail of a word the corpus writes as two)
//! carries `""` in both layers — show nothing for it rather than a neighbour's meaning.
//!
//! See `../../docs/12-word-by-word.md`.

use serde::Deserialize;
use std::collections::HashMap;

#[derive(Debug, Clone, Deserialize, Default)]
pub struct WordByWordPack {
    /// Surah id -> ayahs in id order -> one entry per token.
    pub english: HashMap<String, Vec<Vec<String>>>,
    pub transliteration: HashMap<String, Vec<Vec<String>>>,
}

/// One word of an ayah, in reading order.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GlossedWord {
    /// 1-based index of the word in the ayah.
    pub position: u32,
    /// The ayah's own token.
    pub arabic: String,
    /// The gloss, `""` when the token has none.
    pub english: String,
    pub transliteration: String,
}

/// A hit from the word-level gloss search.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GlossHit {
    pub surah: u32,
    pub ayah: u32,
    pub position: u32,
    pub english: String,
    pub transliteration: String,
}

use crate::Engine;

impl Engine {
    /// Whether a word-by-word pack is loaded at all — cheap enough to gate UI on.
    pub fn word_by_word_loaded(&self) -> bool {
        !self.word_by_word.english.is_empty()
    }

    /// Every word of an ayah, in reading order.
    pub fn words(&self, surah: u32, ayah: u32) -> Vec<GlossedWord> {
        let Some(english) = self.glosses(surah, ayah) else {
            return Vec::new();
        };
        let latin = self.transliterations(surah, ayah).unwrap_or(&[]);
        let tokens: Vec<&str> = self
            .ayah(surah, ayah)
            .map(|a| a.text_arabic.split_whitespace().collect())
            .unwrap_or_default();
        english
            .iter()
            .enumerate()
            .map(|(i, gloss)| GlossedWord {
                position: i as u32 + 1,
                arabic: tokens.get(i).copied().unwrap_or("").to_string(),
                english: gloss.clone(),
                transliteration: latin.get(i).cloned().unwrap_or_default(),
            })
            .collect()
    }

    /// One word, by its 1-based position.
    pub fn word(&self, surah: u32, ayah: u32, position: u32) -> Option<GlossedWord> {
        let words = self.words(surah, ayah);
        if position == 0 {
            return None;
        }
        words.into_iter().nth(position as usize - 1)
    }

    pub fn glosses(&self, surah: u32, ayah: u32) -> Option<&[String]> {
        layer_row(&self.word_by_word.english, surah, ayah)
    }

    pub fn transliterations(&self, surah: u32, ayah: u32) -> Option<&[String]> {
        layer_row(&self.word_by_word.transliteration, surah, ayah)
    }

    /// Ayahs containing a word whose gloss carries `term` — a word-level English search, which
    /// finds ayahs a translation search misses because no translator used that phrasing.
    pub fn find_gloss(&self, term: &str, limit: usize) -> Vec<GlossHit> {
        let needle = term.trim().to_lowercase();
        if needle.is_empty() {
            return Vec::new();
        }
        let mut out = Vec::new();
        // Mushaf order, so the result is stable across runs (HashMap iteration is not).
        for surah in &self.surahs {
            let Some(rows) = self.word_by_word.english.get(&surah.id.to_string()) else {
                continue;
            };
            for (ayah_index, glosses) in rows.iter().enumerate() {
                for (index, gloss) in glosses.iter().enumerate() {
                    if !gloss.to_lowercase().contains(&needle) {
                        continue;
                    }
                    let latin = self
                        .word_by_word
                        .transliteration
                        .get(&surah.id.to_string())
                        .and_then(|rows| rows.get(ayah_index))
                        .and_then(|row| row.get(index))
                        .cloned()
                        .unwrap_or_default();
                    out.push(GlossHit {
                        surah: surah.id,
                        ayah: ayah_index as u32 + 1,
                        position: index as u32 + 1,
                        english: gloss.clone(),
                        transliteration: latin,
                    });
                    if out.len() >= limit {
                        return out;
                    }
                }
            }
        }
        out
    }
}

fn layer_row(layer: &HashMap<String, Vec<Vec<String>>>, surah: u32, ayah: u32) -> Option<&[String]> {
    let rows = layer.get(&surah.to_string())?;
    if ayah == 0 || ayah as usize > rows.len() {
        return None;
    }
    Some(rows[ayah as usize - 1].as_slice())
}
